File: ExamplesCSharp/MapReduce.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace ExamplesCSharp
{
    public static class MapReduceHandsOn
    {
        public static IEnumerable<TResult> MapReduce<TSource, TMapped, TKey, TResult>(
            this IEnumerable<TSource> source,
            Func<TSource, IEnumerable<TMapped>> map, Func<TMapped, TKey> keySelector, 
            Func<IGrouping<TKey, TMapped>, IEnumerable<TResult>> reduce)
        {
            return source.SelectMany(map).GroupBy(keySelector).SelectMany(reduce);
        }

        public static ParallelQuery<TResult> MapReduce<TSource, TMapped, TKey, TResult>(
            this ParallelQuery<TSource> source,
            Func<TSource, IEnumerable<TMapped>> map,
            Func<TMapped, TKey> keySelector,
            Func<IGrouping<TKey, TMapped>, IEnumerable<TResult>> reduce)
        {
            return source.SelectMany(map).GroupBy(keySelector).SelectMany(reduce);
        }

        static char[] delimiters =
            Enumerable.Range(0, 256).Select(i => (char)i)
            .Where(c => Char.IsWhiteSpace(c) || Char.IsPunctuation(c))
            .ToArray();

        public static void Run(string dirPath, string extension)
        {
            var files = 
                    Directory.EnumerateFiles(dirPath, "*" + extension, 
                                             SearchOption.AllDirectories);
            var counts =
                files
                .MapReduce(
                        path => File.ReadLines(path).SelectMany(line => line.Split(delimiters)),
                        word => word,
                        group => new[] { new KeyValuePair<string, int>(group.Key, group.Count()) }
                 );
            foreach (var c in counts.OrderByDescending(kvp => kvp.Value))
            {
                Console.WriteLine($"{c.Key} -> {c.Value}");
            }
        }
    }
}

File: ExamplesFSharp/MapReduce.fs
let mapReduce2 (m : 's -> seq<'m>) (selector : 'm -> 'k) (reduce : 'k*seq<'m> -> seq<'r>) (xs : seq<'s>) =
    xs |> Seq.collect m |> Seq.groupBy selector |> Seq.collect reduce

// need no types in f#
let mapReduce m selector reduce xs =
    xs |> Seq.collect m |> Seq.groupBy selector |> Seq.collect reduce

// make parallel
open FSharp.Collections.ParallelSeq
let mapReduceP m selector reduce xs =
    xs |> PSeq.collect m |> PSeq.groupBy selector |> PSeq.collect reduce

open System.IO
open System.Linq

[<EntryPoint>]
let main argv = 

    let delimiters = 
        Enumerable.Range(0,256) 
        |> Seq.map char
        |> Seq.filter (fun c -> System.Char.IsWhiteSpace c || System.Char.IsPunctuation c)
        |> Seq.toArray

    let files = Directory.EnumerateFiles(argv.[0], "*" + argv.[1], SearchOption.AllDirectories) 

    let split lines = 
        lines 
        |> Seq.collect (fun (l : string) -> l.Split delimiters)

    let counts = 
        files 
        |> mapReduceP (split << File.ReadAllLines) id (fun (k,values) -> [k, Seq.length values])
        |> Seq.sortByDescending (fun (word, count) -> count)

    for c in counts do printfn "%A" c
   
    0 
